src/java/org/apache/cassandra/index/sai/disk/format/Version.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.cassandra.index.sai.disk.v1.V1OnDiskFormat;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * Format version of indexing component, denoted as [major][minor]. Same forward-compatibility rules apply as to
+ * {@link org.apache.cassandra.io.sstable.format.Version}.
+ */
+public class Version
+{
+    // Current version
+    public static final Version BA = new Version("ba", V1OnDiskFormat.instance, (c, i) -> defaultFileNameFormat(c, i, "ba"));
+
+    // These should be added in reverse order so that the latest version is used first. Version matching tests
+    // are more likely to match the latest version, so we want to test that one first.
+    public static final List<Version> ALL = Lists.newArrayList(BA);
+
+    public static final Version EARLIEST = BA;
+    // The latest version can be configured to be an earlier version to support partial upgrades that don't
+    // write newer versions of the on-disk formats.
+    public static final Version LATEST = parse(System.getProperty("cassandra.sai.latest.version", "ba"));
+
+    private final String version;
+    private final OnDiskFormat onDiskFormat;
+    private final FileNameFormatter fileNameFormatter;
+
+    private Version(String version, OnDiskFormat onDiskFormat, FileNameFormatter fileNameFormatter)
+    {
+        this.version = version;
+        this.onDiskFormat = onDiskFormat;
+        this.fileNameFormatter = fileNameFormatter;
+    }
+
+    public static Version parse(String input)
+    {
+        checkArgument(input != null);
+        checkArgument(input.length() == 2);
+        if (input.equals(BA.version))
+            return BA;
+        throw new IllegalArgumentException();
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hashCode(version);
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Version other = (Version)o;
+        return Objects.equal(version, other.version);
+    }
+
+    @Override
+    public String toString()
+    {
+        return version;
+    }
+
+    public boolean onOrAfter(Version other)
+    {
+        return version.compareTo(other.version) >= 0;
+    }
+
+    public OnDiskFormat onDiskFormat()
+    {
+        return onDiskFormat;
+    }
+
+    public FileNameFormatter fileNameFormatter()
+    {
+        return fileNameFormatter;
+    }
+
+    public interface FileNameFormatter
+    {
+        String format(IndexComponent indexComponent, IndexContext indexContext);
+    }
+
+    //
+    // SAI default filename formatter. This is the current SAI on-disk filename format
+    //
+    // Format: <sstable descriptor>-SAI+<version>(+<index name>)+<component name>.db
+    //
+    private static final String SAI_DESCRIPTOR = "SAI";
+    private static final String SAI_SEPARATOR = "+";
+    private static final String EXTENSION = ".db";
```

Comment:
nit: I guess we should move these to the top of the file. (Might as well mention it before someone else does.)

Reply:
Actually, why don't we move it to `Descriptor`? We already have methods in there like `isValidFile()`, which could reuse it.

Reply:
I'm happy moving these up to the head of the class but I'd prefer not to move them into `Descriptor`. I'd prefer that `Descriptor` does it's thing and SAI does it's own thing. The whole idea with formatting was that we could change the naming structure without impacting the SSTable at all.

This may be used in the future for adding a generation to index files so we can, potentially, asynchronously compact the segments, rather than doing that synchronously with the main SSTable compaction.

Reply:
What about just `EXTENSION`?